magento/magento2
# ISSUE-30304 - Mark index removal as non-destructive so indexes can be removed without being present in whitelist json

### Description (*)
Previous state: When an index is present in the database but not in declarative schema, it's funneled into \Magento\Framework\Setup\Declaration\Schema\Operations\DropElement for evaluation. DropElement is considered destructive, so it fails \Magento\Framework\Setup\Declaration\Schema\Diff\Diff::canBeRegistered unless the index is present in a db_schema_whitelist.json file. An index must be present in the whitelist json to be automatically removed during declarative schema processing.

After this PR: The index removal is funneled into a new DropIndex class instead. This is considered non-destructive, so it passes \Magento\Framework\Setup\Declaration\Schema\Diff\Diff::canBeRegistered. Indexes present in the database but not declarative schema are automatically removed, regardless of what's in db_schema_whitelist.json.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#30304

### Manual testing scenarios (*)

1. Directly in MySQL, add an index to a column in a core table.
2. Run bin/magento setup:upgrade.
3. Confirm that the index was automatically removed because it doesn't exist in any declarative schema.

### Questions or comments

I was uncertain if this needs tests written as there seem to be other schema operation classes that aren't tested. If it does need tests, I will need advice on what test class should be updated, and what the test should check for. I understand the concept of unit tests, but haven't written any before.

### Contribution checklist (*)
